privacyidea/lib/policydecorators.py

```diff
@@ -566,10 +562,18 @@ def auth_otppin(wrapped_function, *args, **kwds):
                     return False
 
             if list(otppin_dict)[0] == ACTIONVALUE.USERSTORE:
-                rv = user_object.check_password(pin)
-                return rv is not None
+                # If the PIN should be checked with the userstore, cache the result in options so we only have to do
+                # this once per request.
+                if "otppin_userstore_success" in options and is_true(options["otppin_userstore_success"]):
+                    return True
+                elif "otppin_userstore_success" in options and not is_true(options["otppin_userstore_success"]):
+                    return False
+                else:
+                    rv = user_object.check_password(pin)
+                    options["otppin_userstore_success"] = rv is not None
```

Comment:
This does not work when authentication with PIN+OTP is allowed as well. When we check the "PIN+OTP" against the userstore for a challenge-request, this fails and the `otppin_userstore_success` entry is `False`. When we try the direct authentication later, the (now correctly separated from the OTP) PIN check fails here.
Maybe we can cache the password (hashed) somewhere and compare it before assuming it failed.